### DISCUSSION TITLE
I can't get the built app to run: GitX quit unexpectedly.

### DISCUSSION THREAD

**Discussion by ADTC** [gitx/gitx#366] [Q&A] [answered]
This is a script of commands I'm running to make a build using a MacBook using M1 Pro.

```sh
# Stop on any errors and echo commands. See https://stackoverflow.com/a/2871034/1134080
set -euxo pipefail

# Ensure the correct Xcode app is used (you can cancel script by failing the password entry)
sudo xcode-select -s /Applications/Xcode.app

# Clean up artifacts from last build. See https://git-scm.com/docs/git-clean
git clean -Xdf --exclude=\!/Dev.xcconfig

# Ask to clean up other files (Use `q` to skip clean-up and keep the files)
set +x
echo "If an unwanted file/dir is consistently being asked, add it to .gitignore to clean it automatically."
echo "To skip the prompt for untracked files you want to keep, stage them before running the script."
set -x
git clean -di

# Update submodules (disable this if you're managing submodules yourself)
git submodule sync
git submodule update --init --recursive

# Build the Objective Git dependencies
cd External/objective-git && script/bootstrap && script/update_libgit2 && cd ../..

# Build the GitX app
xcodebuild -workspace GitX.xcworkspace -scheme GitX -archivePath ./GitX archive ARCHS="arm64"

# Run the GitX app (build artifact)
open GitX.xcarchive/Products/Applications/GitX.app
```

I have installed Xcode Version 14.2 (14C18). The `GitX.app` file is generated, but when I run it, I get the following:

<img width="468" alt="image" src="https://user-images.githubusercontent.com/6047296/221375766-e374d1ef-10e3-43ef-8b6d-8f5ed06a675b.png">

With the details here: https://gist.github.com/ADTC/d9a52b702420d5baba732ce05f40fea0

I have tried moving it to the Applications folder, but the issue remains the same.

I have tried using Xcode itself to build and run:

<img width="1140" alt="image" src="https://user-images.githubusercontent.com/6047296/221379302-2bdb325f-01c3-4d69-ae72-a55824e2a344.png">

**What could I be missing?** @hannesa2

**Comment by hannesa2** [comment]
Hmm, normally the CI should show you, how to build it properly

**Reply by ADTC** [reply]
@hannesa2 Unfortunately in this case I'm stuck and don't have a solution yet. I am following the CI steps, and I haven't made any change to the code. Could you review my script above and see if I missed something? (I've updated it with comments.)

Does this ring any bells?

> **Termination Reason:**    Namespace DYLD, Code 1 Library missing
>
> **Library not loaded:** `@loader_path/../Frameworks/MGScopeBar.framework/Versions/A/MGScopeBar`
>
> **Referenced from:** <505F961A-9339-3285-9AB1-9CB2C0C3BC8F> `/Users/USER/*/GitX.app/Contents/MacOS/GitX`
>
> **Reason:**
> * tried: `./GitX.xcarchive/Products/Applications/GitX.app/Contents/MacOS/../Frameworks/MGScopeBar.framework/Versions/A/MGScopeBar` (code signature in <137BBC07-4E2E-370C-BE4A-57FB6D0A7A3C>
> * `./GitX.xcarchive/Products/Applications/GitX.app/Contents/Frameworks/MGScopeBar.framework/Versions/A/MGScopeBar` not valid for use in process: mapped file has no Team ID and is not a platform binary (signed with custom identity or adhoc?))
> * `/System/Volumes/Preboot/Cryptexes/OS@loader_path/../Frameworks/MGScopeBar.framework/Versions/A/MGScopeBar` (no such file)
> * `./GitX.xcarchive/Products/Applications/GitX.app/Contents/MacOS/../Frameworks/MGScopeBar.framework/Versions/A/MGScopeBar` (code signature in <137BBC07-4E2E-370C-BE4A-57F
(terminated at launch; ignore backtrace)

Do I need to build `MGScopeBar` itself somehow? But it's not a step in the CI.

**Comment by hannesa2** [comment]
From the commands I do
```
git clean -xdf
git submodule update --init --recursive
cd External/objective-git && script/bootstrap && script/update_libgit2 && cd ../..
xcodebuild -workspace GitX.xcworkspace -scheme GitX -archivePath ./GitX archive ARCHS="arm64"
```
<img width="1047" alt="image" src="https://user-images.githubusercontent.com/3314607/221401136-59b6f098-ae82-4ca0-8f8f-20d230f99eb4.png">

But when I understand you, you have no build issue, you run into a runtime issue, am I right ?

**Reply by ADTC** [reply]
Yes that's right. It's a runtime issue. What machine did you do the above build in? Is it running properly?

**Reply by hannesa2** [reply]
I use 
<img width="222" alt="image" src="https://user-images.githubusercontent.com/3314607/221405692-ff3c75ec-79fa-412b-9664-4cc30513b4e8.png">

Interesting, my local build run into this too 
<img width="1142" alt="image" src="https://user-images.githubusercontent.com/3314607/221405605-22e4c7ef-6955-4f13-971a-604cb175dc89.png">


But when I use latest master artifact from CI, it works 
<img width="280" alt="image" src="https://user-images.githubusercontent.com/3314607/221405631-3941cf02-953e-4d2a-b57d-07ac1a06c3cf.png">

I've no clue, what's going on here

**Reply by ADTC** [reply]
I've figured it out. Writing my answer below. Please wait.

**Comment by ADTC** [comment, marked as answer]
## OBSOLETE! [Please see README instead.](https://github.com/gitx/gitx#how-to-build-in-xcode) See #376 for the PR fixing this.

This happens because you have not set up your Apple ID and signing certificate properly in the project using Xcode. To resolve this:

1. Open the Xcode app.
2. In Settings > Accounts, if you haven't added your Apple ID yet, click the `+` button and add your Apple ID.
3. In your Apple ID account settings, there should be at least one team with your name and `(Personal Team)` in the name. Click on it.
4. Click on Manage Certificates.
5. If you don't see any certificate listed, click the `+` button and click on Apple Development.
6. Click Done and close the Settings window.
7. In the Project navigator, click on the project root (blue App Store icon with the name `GitX`).
8. In the Signing & Capabilities tab:
   1. Enable "Automatically manage signing" checkbox.
   2. In the Team drop-down, select the same team (your name with `(Personal Team)`).
   3. Your changes are automatically saved.
9. That's it. You can now run either the script above, or you can click the Play button (Product > Run) in Xcode to test.
   - In the first time, your keychain may ask you to authorize the use of the certificate. Enter your local computer account password and click Allow.
   - Tip: It may be a good idea to use Product > Clean Build Folder first before building and running the app again.

# Caution
Don't stage or commit the changes in `GitX.xcodeproj/project.pbxproj` file. Some of these changes are specific for your development environment & Apple ID, and **must not be committed to the repository**.

**Reply by claytonrcarter** [reply]
Does this mean that we need an active Apple dev account and certificate to build the app locally now?

**Reply by ADTC** [reply]
@hannesa2 try this. As for the changes, definitely the `DevelopmentTeam` identifier must not be committed as it is my personal team ID. But I wonder if the rest (such as `CODE_SIGN_IDENTITY` and `CODE_SIGN_STYLE`) is safe to commit or not.

@claytonrcarter it appears so. 😢 I couldn't get it to work without one. (I didn't enroll in the Apple Developer Program btw. I just used my regular Apple ID. Don't worry, you don't need to pay a $99 annual fee to get this working. 😂)

**Reply by claytonrcarter** [reply]
Thanks for digging in to this and being able go get back up and running! Expecting contributors (especially drive by ones like myself) to make these changes but not commit them sounds like a workaround more than a solution, though. I wonder if we can come up with a better solution. I assume this is related to the notarization changes in 726339f or thereabouts.

I have no particular expertise here, either, but I'll see if I can come up with something. @hannesa2 when was the last time that you were able to build and run locally?

**Reply by hannesa2** [reply]
Last time to run local build artifact was during testing the arm64 ~ build https://github.com/gitx/gitx/releases/tag/0.18

**Reply by ADTC** [reply]
@claytonrcarter I can confirm that 726339f is very likely where the need to do this was introduced. I checked out the merge before and after, and tested building without any changes to `GitX.xcodeproj/project.pbxproj`. The merge before, c1a8608, works and runs GitX, but the merge after, 7cc0658, has the crash problem. Introduce the changes back in, and GitX works again.

It doesn't looks like there's any way around it though, because [Hardened Runtime](https://developer.apple.com/documentation/security/hardened_runtime) is needed for the notarization process. It's also probably why you cannot run the app if you didn't sign it properly, and signing it requires you to tie it up with a Development Team ID.

I don't know if we can simply share a common Development Team ID with everyone (including new people forking the repo). It may not be possible because it's tied to an Apple ID - unless we also create and share a common Apple ID with everyone as well. Just thinking out loud here, but definitely this is not ideal.

Looks like this is a problem since Xcode 8 actually. See here; some of these have potential solutions:
* [xcode - How do I not commit the development team lines in project.pbxproj without deselecting those lines manually? - Stack Overflow](https://stackoverflow.com/q/68500460/1134080)
* [ios - How to prevent Xcode 8 from saving "development team" in .pbxproj? - Stack Overflow](https://stackoverflow.com/q/39669661/1134080)
* [ios - ignore project.pbxproj changes related to developer certificates - Stack Overflow](https://stackoverflow.com/q/45754386/1134080)

This could be the easiest solution: https://samwize.com/2020/11/20/using-xcconfig-to-configure-to-your-developer-account/

**Reply by ADTC** [reply]
@claytonrcarter I implemented a solution in #376. I think this will help newcomers get started easily without having to muck up the tracked project files with their private info.

**Comment by hannesa2** [comment]
@ADTC Is this solved ?

**Reply by ADTC** [reply]
I think so. Are you facing a problem still?

**Comment by hannesa2** [comment]
As I understand, you don't run into this error anymore. I've no such issue.
I close it, but please reopen, if it still exists
